src/crimson/osd/main.cc

```diff
@@ -0,0 +1,113 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <iostream>
+
+#include <seastar/core/app-template.hh>
+#include <seastar/core/thread.hh>
+
+#include "common/ceph_argparse.h"
+#include "crimson/common/config_proxy.h"
+#include "crimson/mon/MonClient.h"
+#include "crimson/net/SocketMessenger.h"
+
+#include "osd.h"
+
+using config_t = ceph::common::ConfigProxy;
+
+void usage() {
+  std::cout << "usage: crimson -i <ID>" << std::endl;
+  generic_server_usage();
+}
+
+int main(int argc, char* argv[])
+{
+  std::vector<const char*> args{argv + 1, argv + argc};
+  if (args.empty()) {
+    std::cerr << argv[0] << ": -h or --help for usage" << std::endl;
+    return EXIT_FAILURE;
+  }
+  std::string cluster;
+  std::string conf_file_list;
+  // ceph_argparse_early_args() could _exit(), while local_conf() won't ready
+  // until it's started. so do the boilerplate-settings parsing here.
+  auto init_params = ceph_argparse_early_args(args,
+                                              CEPH_ENTITY_TYPE_CLIENT,
+                                              &cluster,
+                                              &conf_file_list);
+  if (ceph_argparse_need_usage(args)) {
+    usage();
+    return EXIT_SUCCESS;
+  }
+
+  int return_value = 0;
+  // TODO: add heartbeat
+  seastar::app_template app;
+  // talk with osd
+  ceph::net::SocketMessenger cluster_msgr;
+  // talk with mon/mgr
+  ceph::net::SocketMessenger client_msgr;
+  ceph::mon::Client monc{client_msgr};
+  seastar::sharded<OSD> osd;
```

Comment:
@majianpeng i think the `OSD` class is for hosting the shared services like osdmap cache, so probably a better name would be `OSDMapCache` or `OSDService`?

if we will go with 1:1 mapping, we will make OSD a `seastar::sharded<OSD>`. and depending on if we want to share the osdmap cache, we could leave it as a standalone service or just make it a member variable of `OSD` just for better encapsulation instead of for performance consideration.